### PR TITLE
Iox #415 Refactor Service Description Part3: Restrict wildcard usage to searching only

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

This is the third part of a series of pull requests regarding the refactoring of `ServiceDescription`, which paves the way for a new `ServiceRegistry`.

**Done:**
* Add `isValid()` checks in `PortManager` and friends to make sure no invalid `ServiceDescription` is saved into the `ServiceRegistry`
  * Define empty strings as invalid
* Allow wildcards only for searching the `ServiceRegistry` to have well-defined `ServiceDescription`'s

**To be done in follow-up PR as part of #415:**
* Move the`Any_t` from `posh_runtime.hpp` to the new user-facing search discovery header
* [x] Implement `template<class T> cxx::variant(T&& t)` to have a easier user API for searching the `ServiceRegistry` (Done in #864)

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes partly #415

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp [hunk 1]
```diff
@@ -99,27 +92,18 @@ class ServiceDescription
...
-    ServiceDescription(uint16_t f_serviceID, uint16_t f_eventID, uint16_t f_instanceID) noexcept;
+    ServiceDescription(const ServiceDescription&) noexcept = default;
+    ServiceDescription(ServiceDescription&&) noexcept = default;
+    ~ServiceDescription() noexcept = default;
 
     /// @brief construction of the capro service description using fixed strings to create an event service description
-    ServiceDescription(const IdString_t& f_service,
-                       const IdString_t& f_instance,
-                       const IdString_t& f_event,
+    ServiceDescription(const IdString_t& service,
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp:104]: I would love to avoid `isValid()` and invalid `ServiceDescription`'s. However, forbidding the user to provide an empty string is cumbersome and I only came up with those alternatives:
1. Let a factory create all `ServiceDescription`'s, decided against as the user API would get way too complicated
2. Calling the `errorHandler()` if the users passes `{"", "", ""}`, not user friendly, either
3. `static_assert(!service.empty()  &&... )` AFAIK not possible with C++14

File: iceoryx_posh/test/integrationtests/test_roudi_findservice.cpp [hunk 2]
```diff
@@ -41,10 +41,11 @@ class RoudiFindService_test : public RouDi_GTest
 
 TEST_F(RoudiFindService_test, OfferSingleMethodServiceSingleInstance)
 {
-    auto isServiceOffered = senderRuntime->offerService({"service1", "instance1"});
+    auto isServiceOffered = senderRuntime->offerService({"service1", "instance1", "event1"});
     this->InterOpWait();
 
-    auto instanceContainer = receiverRuntime->findService({"service1", "instance1"});
+    auto instanceContainer = receiverRuntime->findService({iox::cxx::in_place_type<IdString_t>(), "service1"},
+                                                          {iox::cxx::in_place_type<IdString_t>(), "instance1"});
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_roudi_findservice.cpp:48]: If there are no objections regarding this interface, I will add a generic c'tor to `cxx::variant` in another PR to simplify the usage to the following:
```
enum class SearchMode
{
    WILDCARD
};

void findService(const std::variant<SearchMode, cxx::string> service, const std::variant<SearchMode, cxx::string> instance)
{
   //.. snip ..
}

int main()
{
    findService("Blubb", SearchMode::WILDCARD);
}
```

This is [supported](https://godbolt.org/z/q3dvnvEMG) by the `std::variant`, [c'tor 4](https://en.cppreference.com/w/cpp/utility/variant/variant) but currently not by our `cxx::variant`.

File: iceoryx_posh/source/capro/service_description.cpp [hunk 3]
```diff
@@ -104,86 +101,38 @@ ServiceDescription::ServiceDescription(const cxx::Serialization& f_serial) noexc
...
-    }
-
-    if (!(cxx::convert::stringIsNumber(m_eventString.c_str(), cxx::convert::NumberType::UNSIGNED_INTEGER)
-          && cxx::convert::fromString(m_eventString.c_str(), m_eventID)))
-    {
-        m_eventID = InvalidID;
-    }
 }
 
 bool ServiceDescription::operator==(const ServiceDescription& rhs) const
```
Comment (mossmaurice) [inline iceoryx_posh/source/capro/service_description.cpp:121]: We could also consider something like this in order to prevent RouDi to connect pub/subs with invalid `ServiceDescription`'s and omit some of the `isValid()` checks. However, I'm not convinced that this makes sense semantically, so I decided against.

```
bool ServiceDescription::operator==(const ServiceDescription& rhs) const
{
    if (!isValid() || !rhs.isValid())
    {
        return false;
    }
    // .. snip ...   
}
```

Comment (mossmaurice) [reply iceoryx_posh/source/capro/service_description.cpp:121]: After a discussion with @MatthiasKillat I'm tempted to implement the above code :grimacing:  @elBoberido What do you think?

File: iceoryx_posh/test/integrationtests/test_roudi_findservice.cpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_posh/test/integrationtests/test_roudi_findservice.cpp:48]: First version [worked](https://github.com/ApexAI/iceoryx/commit/705525d10e28d29b0d4a731c164c44e534e97a15).

File: iceoryx_binding_c/include/iceoryx_binding_c/service_description.h [hunk 4]
```diff
@@ -21,10 +21,6 @@
 
 typedef struct
 {
-    uint16_t serviceId;
-    uint16_t instanceId;
-    uint16_t eventId;
-
```
Comment (elBoberido) [inline iceoryx_binding_c/include/iceoryx_binding_c/service_description.h:27]: Déjà vu. Didn't I see this in the part 1 PR? Please rebase or merge to master

File: iceoryx_posh/source/roudi/port_manager.cpp [hunk 5]
```diff
@@ -151,29 +151,33 @@ void PortManager::handlePublisherPorts() noexcept
...
 {
     publisherPort.tryGetCaProMessage().and_then([this, &publisherPort](auto caproMessage) {
-        m_portIntrospection.reportMessage(caproMessage);
-        if (capro::CaproMessageType::OFFER == caproMessage.m_type)
-        {
-            this->addEntryToServiceRegistry(caproMessage.m_serviceDescription.getServiceIDString(),
-                                            caproMessage.m_serviceDescription.getInstanceIDString());
-        }
-        else if (capro::CaproMessageType::STOP_OFFER == caproMessage.m_type)
+        if (caproMessage.m_serviceDescription.isValid())
```
Comment (mossmaurice) [inline iceoryx_posh/source/roudi/port_manager.cpp:154]: This could be an `Expects`

**Review by elBoberido — COMMENTED** [review]
some test not yet reviewed

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp [hunk 6]
```diff
@@ -44,14 +44,15 @@ class PoshRuntimeImpl : public PoshRuntime
     virtual ~PoshRuntimeImpl() noexcept;
 
     /// @copydoc PoshRuntime::findService
-    cxx::expected<InstanceContainer, FindServiceError> findService(const capro::IdString_t& service,
-                                                                   const capro::IdString_t& instance) noexcept override;
+    cxx::expected<InstanceContainer, FindServiceError>
+    findService(const cxx::variant<SearchMode, capro::IdString_t> service,
+                const cxx::variant<SearchMode, capro::IdString_t> instance) noexcept override;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp:49]: That makes sense. After some thinking I think I prefer `Any`, so:
```
findService(const cxx::variant<Any_t, capro::IdString_t> service,
                   const cxx::variant<Any_t, capro::IdString_t> instance) noexcept override;

```

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp [hunk 7]
```diff
@@ -372,10 +381,33 @@ NodeData* PoshRuntimeImpl::createNode(const NodeProperty& nodeProperty) noexcept
...
+        instanceString = "*";
+    }
+    else
+    {
+        instanceString = *instance.get_at_index<1U>();
+    }
+
     IpcMessage sendBuffer;
-    sendBuffer << IpcMessageTypeToString(IpcMessageType::FIND_SERVICE) << m_appName << service << instance;
+    sendBuffer << IpcMessageTypeToString(IpcMessageType::FIND_SERVICE) << m_appName << serviceString << instanceString;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:410]: That will be a limitation on the current `master`. However, this will be resolved in the next few weeks, so that things will behave just fine for the 2.0 :banana: release.

[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:410]: That's okay, but can you add a todo for this, so that it will no be forgotten

[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:410]: Have a look at line 387 ;)

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp [hunk 8]
```diff
@@ -602,6 +637,31 @@ TEST_F(PortManager_test, AcquireInterfacePortDataAfterDestroyingPreviouslyAcquir
     acquireMaxNumberOfInterfaces(runtimeName);
 }
 
+TEST_F(PortManager_test, DoDiscoveryWithInvalidServiceDescriptionInApplicationPortLeadsToEmptyResponse)
+{
+    auto applicationPortData = m_portManager->acquireApplicationPortData(
+        iox::RuntimeName_t(iox::cxx::TruncateToCapacity, "OhWieSchoenIsPanama"));
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:643]: `TruncateToCapacity` is superfluous

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp [hunk 9]
```diff
@@ -602,6 +637,31 @@ TEST_F(PortManager_test, AcquireInterfacePortDataAfterDestroyingPreviouslyAcquir
...
+    ASSERT_NE(applicationPortData, nullptr);
+
+    auto interfacePortData = m_portManager->acquireInterfacePortData(
+        iox::capro::Interfaces::AMQP, iox::RuntimeName_t(iox::cxx::TruncateToCapacity, "OhWieSchoenIsNicaragua"));
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:647]: same here

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 10]
```diff
@@ -44,13 +44,13 @@ class ServiceRegistry_test : public Test
     iox::roudi::ServiceRegistry registry;
     iox::roudi::ServiceRegistry::InstanceSet_t searchResults;
 
-    iox::cxx::string<100> AnyInstanceString{iox::capro::AnyInstanceString};
+    iox::capro::IdString_t Wildcard{iox::roudi::Wildcard};
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:47]: and here is the requested Wildcard IdString :smile:

File: iceoryx_posh/test/integrationtests/test_roudi_findservice.cpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_roudi_findservice.cpp:48]: In case we have
```cpp
void foo(cxx::variant<int, bool, int> a);
```
What would happens with
```
foo(42);
```
I guess the same as with
```cpp
foo({iox::cxx::in_place_type<int>(), 42})
```

**Review by elBoberido — COMMENTED** [review]
first round finished

File: iceoryx_posh/include/iceoryx_posh/internal/gateway/gateway_generic.inl [hunk 11]
```diff
@@ -79,9 +79,9 @@ GatewayGeneric<channel_t, gateway_t>::addChannel(const capro::ServiceDescription
                                                  const IceoryxPubSubOptions& options) noexcept
 {
     // Filter out wildcard services
-    if (service.getServiceIDString() == capro::IdString_t(cxx::TruncateToCapacity, capro::AnyServiceString)
-        || service.getInstanceIDString() == capro::IdString_t(cxx::TruncateToCapacity, capro::AnyInstanceString)
-        || service.getEventIDString() == capro::IdString_t(cxx::TruncateToCapacity, capro::AnyEventString))
+    if (service.getServiceIDString() == capro::IdString_t(cxx::TruncateToCapacity, "*")
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/internal/gateway/gateway_generic.inl:82]: Well, that's gateway code. Gateways are free to interpret certain strings in a certain way. Let's tackle that in a follow PR.

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp [hunk 12]
```diff
@@ -108,7 +109,11 @@ PoshRuntimeImpl::getMiddlewarePublisher(const capro::ServiceDescription& service
     constexpr uint64_t MAX_HISTORY_CAPACITY =
         PublisherPortUserType::MemberType_t::ChunkSenderData_t::ChunkDistributorDataProperties_t::MAX_HISTORY_CAPACITY;
 
-    /// @todo call service.isValid()
+    if (!service.isValid())
+    {
+        LogWarn() << "Could not create publisher, invalid service description provided";
+        return nullptr;
+    }
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:116]: TBH I prefer here what you said earlier some lines up: "`cxx::Expects` shall not be used when validating user input". I'll leave it like this as the user gets some useful error output.

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp [hunk 13]
```diff
@@ -215,7 +220,11 @@ PoshRuntimeImpl::getMiddlewareSubscriber(const capro::ServiceDescription& servic
 {
     constexpr uint64_t MAX_QUEUE_CAPACITY = SubscriberPortUserType::MemberType_t::ChunkQueueData_t::MAX_CAPACITY;
 
-    /// @todo call service.isValid()
+    if (!service.isValid())
+    {
+        LogWarn() << "Could not create subscriber, invalid service description provided";
+        return nullptr;
+    }
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:227]: See [here](https://github.com/eclipse-iceoryx/iceoryx/pull/854#discussion_r658085718)

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp [hunk 14]
```diff
@@ -423,11 +455,18 @@ bool PoshRuntimeImpl::offerService(const capro::ServiceDescription& serviceDescr
     return false;
 }
 
-void PoshRuntimeImpl::stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept
+bool PoshRuntimeImpl::stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    capro::CaproMessage msg(
-        capro::CaproMessageType::STOP_OFFER, serviceDescription, capro::CaproMessageSubType::SERVICE);
-    m_applicationPort.dispatchCaProMessage(msg);
+    if (serviceDescription.isValid())
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:460]: See [here](https://github.com/eclipse-iceoryx/iceoryx/pull/854#discussion_r658085718)

**Comment by dkroenke** [comment]
@mossmaurice 
Just a short reminder that the API breaking changes in the Service Description (including Part 1 and 2) should be documented in the Changelog.

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp [hunk 15]
```diff
@@ -220,6 +210,25 @@ void setDestroyFlagAndClearContainer(vector& container)
...
+                {iox::capro::InvalidIdString, iox::capro::InvalidIdString, iox::capro::InvalidIdString},
+                publisherOptions,
+                "guiseppe",
+                m_payloadDataSegmentMemoryManager,
+                PortConfigInfo())
+            .value());
+    ASSERT_TRUE(publisher);
+    publisher.offer();
+
+    EXPECT_DEATH({ m_portManager->doDiscovery(); }, ".*");
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:229]: As we're testing `cxx::E`We should revise all death tests, once we implement the release/debug mode behaviour of `cxx::Expects`.

File: iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp [hunk 16]
```diff
@@ -134,10 +130,9 @@ class ServiceDescription
     /// @brief Returns the scope of a ServiceDescription
     Scope getScope() noexcept;
 
-    /// @todo remove any*string from doxygen?
-    ///@brief Returns true for valid ServiceDescription
-    /// false for ServiceDescription that contains either of InvalidID/InvalidIDString  AnyService/AnyServiceString,
-    /// AnyInstance/AnyInstanceString, AnyEvent/AnyEventString.
+    /// @brief Returns true for valid ServiceDescription
+    /// false for ServiceDescription that contain InvalidStrings.
+    /// @return bool, true if ServiceDescription is valid, false otherwise
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp:135]: I see no reason to mention the return type in the brief description.

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp [hunk 17]
```diff
@@ -44,14 +44,15 @@ class PoshRuntimeImpl : public PoshRuntime
     virtual ~PoshRuntimeImpl() noexcept;
 
     /// @copydoc PoshRuntime::findService
-    cxx::expected<InstanceContainer, FindServiceError> findService(const capro::IdString_t& service,
-                                                                   const capro::IdString_t& instance) noexcept override;
+    cxx::expected<InstanceContainer, FindServiceError>
+    findService(const cxx::variant<Any_t, capro::IdString_t> service,
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp:48]: We should be more explicit here I think. Something like `AnyIdString`. I also admit that I would strive not to use the `_t` suffix in types visible to the user, only internal ones. Personally I try to restrict it for low level types (primitive types, pointers, ...) and do not use it for classes.

File: iceoryx_posh/include/iceoryx_posh/runtime/posh_runtime.hpp [hunk 18]
```diff
@@ -51,6 +51,11 @@ enum class FindServiceError
     INSTANCE_CONTAINER_OVERFLOW
 };
 
+/// @brief Used to search for any string (wildcard)
+struct Any_t
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/runtime/posh_runtime.hpp:55]: We should be more explicit and call it `AnyIdString`. Although it is no string, it is used in for the string wildcard search only (e.g. has nothing to do with a general `any` wrapper which can store any object).

File: iceoryx_posh/source/roudi/port_manager.cpp [hunk 19]
```diff
@@ -274,7 +276,7 @@ void PortManager::handleInterfaces() noexcept
         {
             for (auto& instance : x.second.instanceSet)
             {
-                caproMessage.m_serviceDescription = capro::ServiceDescription(x.first, instance, capro::AnyEventString);
+                caproMessage.m_serviceDescription = capro::ServiceDescription(x.first, instance, roudi::Wildcard);
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/roudi/port_manager.cpp:279]: Why not call it `AnyIdString`? What is the scope of use of the `wildcard`? Only for strings?

File: iceoryx_posh/test/moduletests/test_posh_runtime.cpp [hunk 20]
```diff
@@ -146,7 +146,7 @@ TEST_F(PoshRuntime_test, GetMiddlewareApplicationIsSuccessful)
 
     ASSERT_NE(nullptr, applicationPortData);
     EXPECT_EQ(m_runtimeName, applicationPortData->m_runtimeName);
-    EXPECT_EQ(iox::capro::ServiceDescription(), applicationPortData->m_serviceDescription);
+    EXPECT_NE(iox::capro::ServiceDescription(), applicationPortData->m_serviceDescription);
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_posh_runtime.cpp:149]: A not invalid check I presume? The expectation somehow changed, I would think a default constructed `ServiceDescription` is invalid and hence we check that we get something valid? If so, we have the `isValid` method I think.

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp (hunk 15 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:229]: This cannot really happen, can it? We will no request a port with invalid `ServiceDescription` and initialize a publisher with it. Whether discovery should terminate then is debatable, it could just do nothing as well since it cannot offer or subscribe to such ports. But this does not really matter, we should have the system invariant of not providing or managing ports with invalid `ServiceDescription`.

**Review by MatthiasKillat — APPROVED** [review]
Looks fine, small changes to wildcard handling may be necessary in follow-ups.

File: iceoryx_posh/source/capro/service_description.cpp [hunk 21]
```diff
@@ -234,15 +227,10 @@ Interfaces ServiceDescription::getSourceInterface() const noexcept
...
 bool ServiceDescription::isValid() const noexcept
 {
-    return !(m_serviceString == iox::capro::InvalidString
-             || m_serviceString == capro::IdString_t(cxx::TruncateToCapacity, iox::capro::AnyServiceString)
-             || m_instanceString == iox::capro::InvalidString
-             || m_instanceString == capro::IdString_t(cxx::TruncateToCapacity, iox::capro::AnyInstanceString)
-             || m_eventString == iox::capro::InvalidString
-             || m_eventString == capro::IdString_t(cxx::TruncateToCapacity, iox::capro::AnyEventString));
+    return !(m_serviceString == iox::capro::InvalidIdString || m_instanceString == iox::capro::InvalidIdString
+             || m_eventString == iox::capro::InvalidIdString);
```
Comment (elBoberido) [inline iceoryx_posh/source/capro/service_description.cpp:233]: This is just a personal preference and you can keep this as is if you want but I find it always hard to grasp when the result of a longer expression is negated. In this case it would be easier for me to understand if it would be `return (m_serviceString != iox::capro::InvalidIdString && ...)`

File: iceoryx_posh/source/roudi/port_manager.cpp [hunk 22]
```diff
@@ -195,6 +196,7 @@ void PortManager::handleSubscriberPorts() noexcept
 void PortManager::doDiscoveryForSubscriberPort(SubscriberPortType& subscriberPort) noexcept
 {
     subscriberPort.tryGetCaProMessage().and_then([this, &subscriberPort](auto caproMessage) {
+        cxx::Expects(caproMessage.m_serviceDescription.isValid() && "invalid service description!");
```
Comment (elBoberido) [inline iceoryx_posh/source/roudi/port_manager.cpp:199]: pPlease do a check for a valid `ServiceDescription` in `RouDiprocessMessage`. It's not sufficient to do it only in the posh runtime since a user could open the socket and send a manipulated message (same could happen with fuzzing tests). We need to check all parameter for validity when they are de-serialized.

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp (hunk 15 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:229]: Agreed. This code is under our control an therefore we should have only valid `ServiceDescriptions`. In case of an IO boundary, a sanity check has to be performed. Maybe we need two types of `ServiceDescritions`, a user facing one which can be invalid and one which cannot e.g. `ValidatedServiceDescription` which will be created out of the user facing one. We must make more use of the type system to reduce the number of runtime checks.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp (hunk 10 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:48]: since `Wildcard` now is an IdString, this is superfluous ;)

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:227]: see [here](https://github.com/eclipse-iceoryx/iceoryx/pull/854#discussion_r660455240) ;)

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:460]: see [here](https://github.com/eclipse-iceoryx/iceoryx/pull/854#discussion_r660455240) ;)

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:116]: Makes sense, I mixed up the lines. Was talking about `{offer,stop}Service()`.

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:460]: Fixed

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp (hunk 15 above)
Comment (mossmaurice) [reply iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:229]: TBH I don't like and disagree with statements "This code is under control". With the next refactoring in v3.0 other developers might work on this part of the code and there might be the case where erroneous code slips in and an invalid will make it to the `PortManager`. My goal when developing safety software is always "Better safe than sorry". Additionally, we need to test **all** possible branches due to MC/DC anyway. Hence, I reject the changes and instead we can allow users to disable checking `cxx::Expects` during runtime e.g. in Release mode.

File: iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp (hunk 16 above)
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp:135]: It doesn't hurt either ;)

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp (hunk 17 above)
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp:48]: That's sound like something we should finally bring on the table in the next developer meetup. We are still using various different naming conventions in iceoryx.

File: iceoryx_posh/include/iceoryx_posh/runtime/posh_runtime.hpp (hunk 18 above)
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/runtime/posh_runtime.hpp:55]: You're right there. However, I'd rather change it back to `Wildcard`. `AnyIdString` is too cryptic for me personally.

File: CHANGELOG.md [hunk 23]
```diff
@@ -90,6 +90,18 @@ iox::posix::posixCall(unlink)(sockAddrPublisher.sun_path)
     });
 ```
 
+Service Description is only string-based and well defined:
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply CHANGELOG.md:93]: Added. The default value behaviour hasn't changed. Just the underlying value.

File: iceoryx_posh/source/capro/service_description.cpp (hunk 21 above)
Comment (mossmaurice) [reply iceoryx_posh/source/capro/service_description.cpp:233]: Alrighty, for me it's the other way round :upside_down_face:

File: iceoryx_posh/source/roudi/port_manager.cpp (hunk 22 above)
Comment (mossmaurice) [reply iceoryx_posh/source/roudi/port_manager.cpp:199]: Correct, that's important. Added.

**Review by MatthiasKillat — DISMISSED** [review]

File: iceoryx_posh/source/roudi/roudi.cpp [hunk 24]
```diff
@@ -234,6 +234,8 @@ void RouDi::processMessage(const runtime::IpcMessage& message,
             capro::ServiceDescription service(cxx::Serialization(message.getElementAtIndex(2)));
             cxx::Serialization portConfigInfoSerialization(message.getElementAtIndex(7));
 
+            cxx::Expects(service.isValid() && "invalid service description!");
```
Comment (elBoberido) [inline iceoryx_posh/source/roudi/roudi.cpp:237]: Okay, this is exactly the place where we shouldn't use `cxx::Expects`. It's an input from a user application we potentially cannot trust and we need to assume that it is indeed invalid. Please replace it with a check similar to three lines below

File: iceoryx_posh/source/roudi/roudi.cpp [hunk 25]
```diff
@@ -280,6 +282,7 @@ void RouDi::processMessage(const runtime::IpcMessage& message,
             capro::ServiceDescription service(cxx::Serialization(message.getElementAtIndex(2)));
             cxx::Serialization portConfigInfoSerialization(message.getElementAtIndex(8));
 
+            cxx::Expects(service.isValid() && "invalid service description!");
```
Comment (elBoberido) [inline iceoryx_posh/source/roudi/roudi.cpp:285]: same here

File: CHANGELOG.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply CHANGELOG.md:93]: It is still not fully clear. Pleas state in the note that the wildcards are gone and what does well defined mean in this context?

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp (hunk 15 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:229]: Just to be clear. I didn't suggest to remove the check in the `PortManager`. This is an invariant which needs to be upheld since it is theoretically possible to have an invalid `ServiceDescription`. This is only because the handling with an `cxx::optional` is quite cumbersome in this part of the user interface. But this is again a perfect example why we shouldn't have nullable types. We have to expect them to be null everywhere they are used.

But this test still does not make sense. There should never be a case where `doDiscovery` accesses an invalid `ServiceDescription` since it is not a parameter of the method. It must already be caught in `acquirePublisherPortData` since this is the place where the invariant is broken. This is what I meant with `the code is under our control`.

Agree with you that such a split of a potentially invalid user-facing `ServiceDescription` and a proven always valid internal `ServiceDescription` would be something for another PR.

Oh, and you need to code a bit in Rust. It will make you appreciate the type system and offers you new perspectives to runtime checks ;)

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:452]: Did you maybe forget to push the changes or do I need some glasses :laughing:

File: iceoryx_posh/source/roudi/roudi.cpp (hunk 24 above)
Comment (mossmaurice) [reply iceoryx_posh/source/roudi/roudi.cpp:237]: :see_no_evil: Of course you're right. Fixed.

File: iceoryx_posh/source/roudi/roudi.cpp (hunk 25 above)
Comment (mossmaurice) [reply iceoryx_posh/source/roudi/roudi.cpp:285]: Done

File: CHANGELOG.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (mossmaurice) [reply CHANGELOG.md]: Added another sentence.

File: iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp (hunk 15 above)
Comment (mossmaurice) [reply iceoryx_posh/test/moduletests/test_roudi_portmanager.cpp:229]: Understood. Adapted the test accordingly.

> Oh, and you need to code a bit in Rust. It will make you appreciate the type system and offers you new perspectives to runtime checks ;)

For sure :crab:

File: iceoryx_posh/source/runtime/posh_runtime_impl.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/posh_runtime_impl.cpp:452]: Seems like I need some new :eyeglasses: :smile: This method will be moved to the new user-facing `ServiceDiscovery` API and therefore shall not just terminate in case of wrong user input, but instead inform the user on what went wrong. Hence, no change here.

File: iceoryx_posh/source/roudi/port_manager.cpp [hunk 26]
```diff
@@ -151,6 +151,7 @@ void PortManager::handlePublisherPorts() noexcept
 void PortManager::doDiscoveryForPublisherPort(PublisherPortRouDiType& publisherPort) noexcept
 {
     publisherPort.tryGetCaProMessage().and_then([this, &publisherPort](auto caproMessage) {
+        cxx::Expects(caproMessage.m_serviceDescription.isValid() && "invalid service description!");
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/roudi/port_manager.cpp:154]: We definitely want something more modern to print output from Expects. Does not belong in this PR though, I just dislike this style.

**Review by MatthiasKillat — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]
